RegisterApplication with object config

Sauloxd opened the pull request:
## What this PR does?
- Enhance current registerApplication API to use object configuration. (#475 )
- Created a new error to help users to identify invalid keys in runtime (If this PR is approved, I'll open a PR to S-SPA blog to add this new error)
- Tried to separate internals to handle obj configuration vs arguments (I'm not sure this is the best approach)
- Updated our TS definitions to enable uses to pass the first argument as an Object
   - I couldn't find a way to not make 2nd, 3rd, and 4th args optional! Since if the user uses the object configuration, the next arguments would be unnecessary (not even optional, but invalid)
- Added some tests to use both object and arguments api

## What it currently doesn't do (yet). I will do it after some feedback from you guys
- Improve isActive to accept prefix, Array of prefixes and/or ActivityFn
- Improva registerApplication to accept `domElement` (I'm sorry, I couldn't even find in the repo where this domElement is used by! I only found  reference in helpers libs like `single-spa-parcel` [here](https://github.com/single-spa/single-spa-react/blob/master/src/single-spa-react.js#L150)

Sauloxd commented on spec/apis/register-application.spec.js:
```diff
@@ -19,7 +19,7 @@ describe("registerApplication", function() {
     it(`should throw an error if the name isn't a non empty string`, () => {
       expect(() => {
-        singleSpa.registerApplication(app);
+        singleSpa.registerApplication(null);
```
Changed usage of `app` as a wrong argument, because object arguments triggers object configuration validations

Sauloxd commented on src/applications/app.helpers.js:
```diff
@@ -74,3 +74,111 @@ export function isParcel(appOrParcel) {
+
+  if (
+    !!customProps &&
+    (typeof customProps !== "object" || Array.isArray(customProps))
```
Changed validation for `customProps` because, since it's optional, it should skip the validation if it's not present

Sauloxd commented on src/applications/app.helpers.js:
```diff
@@ -74,3 +74,111 @@ export function isParcel(appOrParcel) {
+  if (invalidPassedKeys.length !== 0)
+    throw Error(
+      formatErrorMessage(
+        30,
```
I peeked [here](https://github.com/single-spa/single-spa.js.org/tree/f24a828c617ac78779dc11a2ac11d471a62f607f/website/src/components/Errors/Codes), and 30 was  the next valid errorCode

Sauloxd commented on src/applications/app.helpers.js:
```diff
@@ -74,3 +74,111 @@ export function isParcel(appOrParcel) {
+
+  validateArgumentsTypeAPI(
+    configurationObject.name,
+    configurationObject.load,
```
Yeah, I'm no fan of the `load` prop either!
In my opinion, having both load/app would cause confusion, since it makes possible for wrongly defining both `load` and `app` in the same registerApplication. Having only 1 way to pass an App I think it's best!
`app` should be more friendly and obvious, since the term `Application` is being used in the documentation

Sauloxd commented on src/applications/app.helpers.js:
```diff
@@ -74,3 +74,111 @@ export function isParcel(appOrParcel) {
+  validateArgumentsTypeAPI(
+    configurationObject.name,
+    configurationObject.load,
+    configurationObject.isActive,
```
Yeah, I was thinking if sharing validations for both object and argument would be the best approach to reduce code duplication, but it would only work if they really shared the same interface, which they won't in the near future...
Since object configuration will grow more complex overtime, I'll try to isolate both validations, specific for argument and object!

Sauloxd commented on src/applications/apps.js:
```diff
@@ -41,61 +42,21 @@ export function getAppStatus(appName) {
-      formatErrorMessage(22, __DEV__ && "customProps must be an object")
-    );
+export function registerApplication(...args) {
+  const { name, loadImpl, activityFn, customProps } = sanitizeAPI(args);
```
I agree, the option 1 would definitely be better for a project that should be as slim as possible, I will try to keep that in mind next time!
My thought process was that naming `appName` to `appNameOrConfig` would make things go a little bit out of hand, and to avoid clutter inside the `registerApplication` function, I moved all arguments to the "sanitize" (where they are being validated and "transformed"  to correct internal api types, well almost hehehe)
If possible I'd like to keep the flow:
`registerApplication` -> `sanitize()` (which will validate all arguments, regarding form, and map outside arguments to correct internal representation) -> apps.push()

Sauloxd commented on src/applications/apps.js:
```diff
@@ -41,61 +42,21 @@ export function getAppStatus(appName) {
       )
     );
 
   apps.push({
```
I like this!

Sauloxd commented:
> Question
Are you planning on creating a PR after this one to add the new behavior for strings-as-activity functions and arrays of activity functions? If so, you can ignore my comments for now about implementing that PR in this PR.

Since I moved few thinks and was adding a few changes on how validations of registerApp was made, I thought it would be best to move slow and checking with you guys if I was on the correct path! But If it's better for you, I could do all in this PR, no problem!

> Let's decide on an interface
We have isActive, activeWhen, and activityFn. Also, we have load, app, and loadImpl. We should arrive at a single terminology if possible.

`isActive, activeWhen, and activityFn`
Hmm... Suffixing with `Fn` seems kinda lazy, as majority of JS devs uses VSCode, which will help them know it should/can be a function with TS.
**isActive** seems like a boolean in name (to me at least), but it's exactly how `react-router` [uses](https://reacttraining.com/react-router/web/api/NavLink/isactive-func), so it should have at least some approval in the community
**activeWhen** is ok too

`load, app, and loadImpl`
Internally loadImpl was being used, so I didn't dare to change it... But I vote on `app`, since using "loadSomething" would imply in async function which will get an app bundle. It is the main use-case for Single-SPA though, so thats a valid point too.
Also, what about changing internally the name from "loadImpl"  to "loadApp"?

obsj: (what is Impl? is it implementation?)

>Tests 

Yes, yes and yes.
Sadly I'm not sure on what things should be best to test, so I'll try to improve on that! thanks

joeldenning commented:
> Since I moved few thinks and was adding a few changes on how validations of registerApp was made, I thought it would be best to move slow and checking with you guys if I was on the correct path! But If it's better for you, I could do all in this PR, no problem!

I appreciate that - makes it easier for us to review it. Let's keep this PR to adding the object version, *but not the new functionality with isActive strings*. We can do that one in a follow-up PR.

> Hmm... Suffixing with Fn seems kinda lazy, as majority of JS devs uses VSCode, which will help them know it should/can be a function with TS.
isActive seems like a boolean in name (to me at least), but it's exactly how react-router uses, so it should have at least some approval in the community
activeWhen is ok too

Good point that `isActive` sounds like a boolean. Maybe let's rename everything to `activeWhen`.

> Internally loadImpl was being used, so I didn't dare to change it... But I vote on app, since using "loadSomething" would imply in async function which will get an app bundle. It is the main use-case for Single-SPA though, so thats a valid point too.

The tests will catch things if you rename `loadImpl` -- feel free to do so. I'd prefer arriving at a single terminology if possible.

> Also, what about changing internally the name from "loadImpl" to "loadApp"?

Sure, I like it.

>  Sadly I'm not sure on what things should be best to test, so I'll try to improve on that! thanks

Your tests so far as good - once you add the couple more I suggested I think the tests will be pretty comprehensive.

Sauloxd commented:
@joeldenning Updated PR with things we discussed!
- normalized terminology
- updated tests to stop rewriting testing but to add new ones with object config flow
- Improve sanitizeArguments to output the partial interface of an `app` internal representation
- use specific validation flows for simpler api vs object api
- pending errors to be add on documentation -> 30, 31
- renamed a few error messages trying to be more consistent

joeldenning commented on src/applications/apps.js:
```diff
@@ -228,3 +201,172 @@ function immediatelyUnloadApp(app, resolve, reject) {
+      formatErrorMessage(
+        20,
+        __DEV__ &&
+          "The config.app on registerApplication must be an application or a loading fuinction"
```
```suggestion
          "The config.app on registerApplication must be an application or a loading function"
```

joeldenning commented on src/applications/apps.js:
```diff
@@ -228,3 +201,172 @@ function immediatelyUnloadApp(app, resolve, reject) {
+      )
+    );
+
+  // @TODO this is not yet implemented
```
Let's save this for your next PR, and not include it in this one. You can move this to a different branch for now and then once this PR is merged you can use that branch for your next PR.

joeldenning commented on typings/single-spa.d.ts:
```diff
@@ -65,18 +65,31 @@ declare module "single-spa" {
-          config: T & AppProps
-        ) => Promise<LifeCycles<T> | Splat<LifeCycles<T>>>),
-    activityFn: (location: Location) => boolean,
+    appNameOrObjectConfiguration: string,
```
Since this is an overloaded function, we don't need to worry about this first definition ever being used when the first argument is a configuration object.

```suggestion
    appName: string,
```

Sauloxd commented:
Updated:
 - Fix typos
 - Revert check for new API on `activeWhen`
 - Removed tests for new API on `activeWhen`
 - Fixed conflicts

Sauloxd commented:
Yes! I’ll work on enhancing activeWhen next
